aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,951 @@
+            .get_global_hpke_keypairs()
+            .inspect(|keypairs| debug!(?keypairs, "table state before running key rotator"))
+            .await?
+            .into_iter()
+            .map(|keypair| (*keypair.id(), keypair))
+            .collect();
+
+        let ids: HashSet<_> = keypairs.iter().map(|(&id, _)| u8::from(id)).collect();
+        // Try to cycle through the entire u8 space before going back to zero. This allows us a
+        // bigger window to quicky reject reports from faulty old clients with an outdated HPKE
+        // config error, rather than attempting to decrypt them with the wrong key.
+        let newest_id = keypairs
+            .iter()
+            .max_by_key(|(&id, _)| id)
+            .map(|(&id, _)| id.into())
+            .unwrap_or(0);
+        let available_ids = Box::new((newest_id..=u8::MAX).chain(0..newest_id).filter_map(
+            move |id| {
+                if !ids.contains(&id) {
+                    Some(HpkeConfigId::from(id))
+                } else {
+                    None
+                }
+            },
+        ));
+
+        Ok(Self {
+            clock: tx.clock().clone(),
+            initially_empty: keypairs.is_empty(),
+            keypairs,
+            available_ids,
+            config,
+        })
+    }
+
+    /// Bootstrap policy:
+    ///   - If this is the very first run of Janus, insert keypairs in [`HpkeKeyState::Active`]
+    ///     for each configured ciphersuite. Detect first run by whether the `global_hpke_keys`
+    ///     table is empty.
+    ///   - For each configured ciphersuite, if there is no pending or active key, insert
+    ///     a key in [`HpkeKeyState::Pending`].
+    fn bootstrap(mut self) -> Result<Self, DatastoreError> {
+        for ciphersuite in &self.config.ciphersuites {
+            if self.initially_empty {
+                info!(?ciphersuite, "bootstrapping new database");
+                self.put(*ciphersuite, HpkeKeyState::Active)?;
+            } else if !self.keypairs.values().any(|keypair| {
+                &keypair.ciphersuite() == ciphersuite
+                    && (keypair.is_active() || keypair.is_pending())
+            }) {
+                info!(?ciphersuite, "bootstrapping new configuration");
+                self.put(*ciphersuite, HpkeKeyState::Pending)?;
+            }
+        }
+        Ok(self)
+    }
+
+    /// Rotation policy:
+    ///   - If the key has been in [`HpkeKeyState::Expired`] for at least
+    ///     [`HpkeKeyRotatorConfig::expired_duration`], delete it.
+    ///   - If the key has a ciphersuite that is not known, safely phase it out. If the key is
+    ///     in [`HpkeKeyState::Pending`], delete it immediately. If it is in
+    ///     [`HpkeKeyState::Active`], move it to expired.
+    ///   - For each configured ciphersuite, move the latest pending key to active if it has been
+    ///     pending for at least [`HpkeKeyRotatorConfig::pending_duration`]. Delete other pending
```

Comment:
> though this isn't guaranteed if an operator manually adds one.

Indeed that's the case I'm trying to cover. If the operator is adding a new pending key, and there is already one present, both could be promoted at the same time, and one of them will be immediately expired. Which one is expired is indeterminate, since they'll have the same update timestamp--or rather there's no guarantee that it'll be the one that the operator didn't insert.

Note the window for the "both promoted at the same time" case is not gated by `pending_duration`, but by how often the key rotator is run. If it's run once per day/week/etc., the window for this case during a manual key rotation is fairly large.

I figured this was an easy enough logic add-on to simplify manual key rotation.